src/libutil/strings.cc

```diff
@@ -48,4 +48,69 @@ template std::string dropEmptyInitThenConcatStringsSep(std::string_view, const s
 template std::string dropEmptyInitThenConcatStringsSep(std::string_view, const std::set<std::string> &);
 template std::string dropEmptyInitThenConcatStringsSep(std::string_view, const std::vector<std::string> &);
 
+/**
+ * Shell split string: split a string into shell arguments, respecting quotes and backslashes.
+ *
+ * Used for NIX_SSHOPTS handling, which previously used `tokenizeString` and was broken by
+ * Arguments that need to be passed to ssh with spaces in them.
+ */
+std::list<std::string> shellSplitString(std::string_view s)
+{
+    std::list<std::string> result;
+    std::string current;
+    bool inQuoteSingle = false;
+    bool inQuoteDouble = false;
+    bool escaped = false;
+    bool lastCharWasQuote = false;
+    for (char c : s) {
+        lastCharWasQuote = false;
+        if (escaped) {
```

Comment:
Maybe this should be disabled while in a single quoted string, since
```
$ echo "\\"
\

$ echo '\\'
\\
```

The logic of this function is quite hard to follow. Maybe it should be something like:
```
auto pop = [&]() -> char {
    auto c = s.front();
    s.remove_prefix(1);
    return s;
}

while (!s.empty()) {
    auto c = pop();
    if (c == '\'') {
        ... copy characters until we encounter another '
    } else if (c == '"') {
        ... copy and unescape characters until we encounter another "
    } else if (c == '\\') {
        if (!s.empty())
            current.push_back(pop());
    } else if (c is whitespace) {
        result.push_back(current);
        current.clear();        
    } else {
        current.push_back(c);
    }
}
```
or something along those lines.

Reply:
The current implementation has been adjusted to look like this